### DISCUSSION TITLE
Filtering question

### DISCUSSION THREAD

**Discussion by the1gofer** [sqlalchemy/sqlalchemy#6695] [Usage Questions] [answered]
Hello,

I have a two models I am trying to work with.    

URL which has a property scraped.
Source which has a property URL.

I would like to find all of the sources that have scraped urls set to True.  This code:

`session.query(Source).filter(Source.url.scraped == True).all()`

give me an error 
'AttributeError: Neither 'InstrumentedAttribute' object nor 'Comparator' object associated with Source.url has an attribute 'scraped'

How would I go about this?   probably use list comprehension, but I feel like their should be a more direct method.

**Comment by zzzeek** [comment, marked as answer]
hey there -

SQLAlchemy wants you to think in terms of a SQL statement.  In this case you are selecting rows from a "source" table, but then you want to filter those rows based on rows in another table "url".  In SQL we use a [join](https://docs.sqlalchemy.org/en/14/tutorial/data_select.html#explicit-from-clauses-and-joins) to achieve this, so SQL wise you want to see:

```sql
   SELECT source.id, source.x, source.y, ... FROM source JOIN url ON source.id=url.source_id WHERE url.scraped == 1
```
the ORM Query object (and the select()) object support this using the [join](https://docs.sqlalchemy.org/en/14/orm/query.html?highlight=query%20join#sqlalchemy.orm.Query.join) method:

```py
    session.query(Source).join(Source.url).filter(URL.scraped == True).all()
```

Another option that creates less efficient SQL is the relationship operator [any()](https://docs.sqlalchemy.org/en/14/orm/internals.html?highlight=relationship%20any#sqlalchemy.orm.RelationshipProperty.Comparator.any) - this will use a correlated subquery with the EXISTS keyword instead of a JOIN:

```py
   session.query(Source).filter(Source.url.any(scraped=True)).all()
```

**Reply by godmar** [reply]
We are facing a similar problem, except that we are trying to add filters programmatically (that is, not hardcoding a specific table name). 

In your example, is it possible to obtain a reference to the `URL` (model) table from the `Source.url` InstrumentedAttribute?

We'd like to translate `Source.url.scraped` into something like
```python
a = Source
b = getattr(a, "url")
c = mystery_step_were_missing(b)
d = getattr(c, "scraped")
```
What is the step to get the `Table` from the InstrumentedAttribute?  If not the Table object itself, can we get at least the table name (and then look it up via metadata?)

**Reply by zzzeek** [reply]
The Source.url attribute will refer to ``Source.url.mapper`` that gives you the target [mapper](https://docs.sqlalchemy.org/en/14/orm/mapping_api.html?highlight=mapper#sqlalchemy.orm.Mapper).  Source.url.property is also the [RelationshipProperty](https://docs.sqlalchemy.org/en/14/orm/internals.html?highlight=relationshipproperty#sqlalchemy.orm.RelationshipProperty) which has ".parent" and ".mapper".
